### PR TITLE
refactor: deactivate caching if a component is part of a cycle

### PR DESCRIPTION
**Pull request by ogabrielluiz**
This pull request adds cycle detection and caching functionality to the `Vertex` class. It introduces the `has_cycle_edges` method to check if a vertex has any cycle edges. Additionally, it includes the `apply_on_outputs` method to apply functions to the outputs of a vertex. The pull request also adds a utility function, `find_cycle_vertices`, to find all vertices that are part of cycles in a directed graph. The `instantiate_component` method is updated to use the `initialize.loading.instantiate_class` function for custom component instantiation. Unit tests are added for the `find_cycle_vertices` utility function. The `set_cache_to_vertices_in_cycle` method is introduced to enable caching for vertices involved in cycles. The vertex instantiation process is refactored for better code organization.

### CODE REVIEW

File: src/backend/base/langflow/graph/graph/base.py [hunk 1]
```diff
@@ -1632,7 +1657,7 @@ def build_edge(self, edge: EdgeData) -> CycleEdge | Edge:
             raise ValueError(f"Source vertex {edge['source']} not found")
         if target is None:
             raise ValueError(f"Target vertex {edge['target']} not found")
-        if (source.id, target.id) in self.cycles:
+        if any(v in self.cycle_vertices for v in [source.id, target.id]):
```
Comment (ogabrielluiz) [inline src/backend/base/langflow/graph/graph/base.py:1660]: @phact  this is what made things start working.

File: src/backend/tests/unit/graph/graph/test_cycles.py [hunk 2]
```diff
@@ -109,3 +114,99 @@ def test_cycle_in_graph_max_iterations():
...
+    non_cycle_outputs_lists = [
+        vertex._custom_component.outputs for vertex in graph.vertices if vertex.id not in cycle_vertices
+    ]
+    non_cycle_outputs = [output for outputs in non_cycle_outputs_lists for output in outputs]
+    for output in non_cycle_outputs:
+        assert output.cache is True
+
+
+@pytest.mark.api_key_required
+def test_updated_graph_with_prompts():
```
Comment (ogabrielluiz) [inline src/backend/tests/unit/graph/graph/test_cycles.py:152]: @phact  Check out this test to understand a bit more what is happening.

File: src/backend/tests/unit/graph/graph/test_cycles.py [hunk 3]
```diff
@@ -109,3 +114,99 @@ def test_cycle_in_graph_max_iterations():
...
+        match_text="testtesttesttest",
+        operator="equals",
+        message=concat_component.concatenate,
+    )
+    text_output = TextOutputComponent(_id="text_output")
+    text_output.set(input_value=router.true_response)
+    chat_output = ChatOutput(_id="chat_output")
+    chat_output.set(input_value=text_output.text_response)
+
+    graph = Graph(chat_input, chat_output)
```
Comment (phact) [inline src/backend/tests/unit/graph/graph/test_cycles.py:136]: Not specific to this PR but how come you need to pass an end component to Graph? Wouldn't it eventually get added through the adjacency lists?

Comment (ogabrielluiz) [reply src/backend/tests/unit/graph/graph/test_cycles.py:136]: Yeah. This is kinda weird still. We went for a syntax similar to Keras' Functional API. 
The end shouldn't be necessary but it has its uses still, I just didn't mind going over it yet.

File: src/backend/tests/unit/graph/graph/test_cycles.py [hunk 4]
```diff
@@ -109,3 +114,99 @@ def test_cycle_in_graph_max_iterations():
...
+
+    # Assertions for graph cyclicity and correctness
+    assert graph.is_cyclic is True, "Graph should contain cycles."
+
+    # Run and validate the execution of the graph
+    results = []
+    max_iterations = 20
+    snapshots = [graph.get_snapshot()]
+
+    for result in graph.start(max_iterations=max_iterations, config={"output": {"cache": False}}):
```
Comment (phact) [inline src/backend/tests/unit/graph/graph/test_cycles.py:203]: We probably want to make max_iterations required if the graph is cyclic?

Something like this in Graph.start()?
```
        if self.is_cyclic:
            if max_iterations is None:
                raise ValueError("max_iterations must be provided for cyclic graphs")
```

would we set other validations for max_iterations? > 0 and maybe a max value?

Comment (ogabrielluiz) [reply src/backend/tests/unit/graph/graph/test_cycles.py:203]: Done. Thanks!

**Review by phact — APPROVED** [review]

**Pull request merged**